### PR TITLE
Functional Tests for Tracer Perp

### PR DESCRIPTION
**Pull request by raymogg**
# Motivation
In order to catch further bugs, some functional tests are being introduced.

# Changes
- adds functional tests
- fixes bug where time was not initialised in `pricing.sol`
- fixes bug with `getTWAPs` function in `pricing.sol`
- fixes bug with ordering in `recordTrade` function in `pricing.sol`

### CODE REVIEW

File: deploy/FullDeploy.js [hunk 1]
```diff
@@ -103,6 +103,13 @@ module.exports = async function (hre) {
         contract: "OracleAdapter",
     })
 
+    const gasOracleAdapter = await deploy("GasOracleAdapter", {
+        from: deployer,
+        log: true,
+        args: [gasOracle.address],
+        contract: "OracleAdapter",
+    })
+
```
Comment (CalabashSquash) [inline deploy/FullDeploy.js]: Should this be being used in the deployment?

**Review by sporejack — COMMENTED** [review]
LGTM!

**Review by sporejack — APPROVED** [review]
LGTM!

**Review by CalabashSquash — CHANGES_REQUESTED** [review]

File: deploy/FullDeploy.js (hunk 1 above)
Comment (raymogg) [reply deploy/FullDeploy.js]: ah I was going to add it in but decided that was most likely a separate PR as there is some logic to go into the perp itself. 

I'll remove this for now and add it in another PR

**Review by CalabashSquash — APPROVED** [review]

**Pull request merged**